# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2024-03-28T19:00:51.443139294+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1e9wbuap/model/ with prefix 4f6515c23f9c4746
[INFO 2024-03-28T19:00:51.455203488+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:00:51.455481851+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2024-03-28T19:00:56.424591062+00:00 kernel.cc:1214] Loading model from path /tmp/tmprq5rfqq9/model/ with prefix 243996bbb943479a
[INFO 2024-03-28T19:00:56.435400936+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2024-03-28T19:00:56.435476754+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmpkluco11t as temporary training directory


[INFO 2024-03-28T19:03:10.939432004+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkluco11t/model/ with prefix 757139bb2e194a90
[INFO 2024-03-28T19:03:10.955455578+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2024-03-28T19:03:10.955505831+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2024-03-28T19:03:10.955535729+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2024-03-28T19:03:21.461365548+00:00 kernel.cc:1214] Loading model from path /tmp/tmpanjw51n3/model/ with prefix bf75789721c64239
[INFO 2024-03-28T19:03:21.466451922+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:03:21.466514164+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2024-03-28T19:03:23.118654846+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmap1l9be/model/ with prefix 42242dd816994c93
[INFO 2024-03-28T19:03:23.140679198+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2024-03-28T19:03:24.359036+00:00 kernel.cc:1214] Loading model from path /tmp/tmppbb48mz0/model/ with prefix 7320d3cfde374fc5
[INFO 2024-03-28T19:03:24.364004425+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2024-03-28T19:03:26.796077011+00:00 kernel.cc:1214] Loading model from path /tmp/tmptxbrppvv/model/ with prefix 22399d02c48f46dd
[INFO 2024-03-28T19:03:26.824622683+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2024-03-28T19:03:29.163453257+00:00 kernel.cc:1214] Loading model from path /tmp/tmplyy0nksa/model/ with prefix f78f066400e44551
[INFO 2024-03-28T19:03:29.169492669+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2024-03-28T19:03:30.2828672+00:00 kernel.cc:1214] Loading model from path /tmp/tmp52gbsf8p/model/ with prefix 862007d9e83d4555
[INFO 2024-03-28T19:03:30.286044403+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2024-03-28T19:03:31.571419583+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcvlo8b2w/model/ with prefix 2b59dc987dbf47ba
[INFO 2024-03-28T19:03:31.579568153+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:03:31.57961435+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2024-03-28T19:03:33.270652137+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkzhdyr3n/model/ with prefix 1141197d582d47d1
[INFO 2024-03-28T19:03:33.292697894+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2024-03-28T19:03:34.597435486+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9b5olg4n/model/ with prefix a9965d31b7fa4bb0
[INFO 2024-03-28T19:03:34.608175057+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2024-03-28T19:03:36.662393865+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4pt3bzd7/model/ with prefix 73298fb4ee194777
[INFO 2024-03-28T19:03:36.682439857+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2024-03-28T19:03:38.203864068+00:00 kernel.cc:1214] Loading model from path /tmp/tmptn3p0vz1/model/ with prefix 00183347578c457f
[INFO 2024-03-28T19:03:38.21290453+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2024-03-28T19:03:39.750660893+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7hovl9t6/model/ with prefix e93ab76e2a2d46c7
[INFO 2024-03-28T19:03:39.766759747+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2024-03-28T19:03:41.000696047+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6w4df2m3/model/ with prefix 5afa30965a9242de
[INFO 2024-03-28T19:03:41.007482447+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2024-03-28T19:03:42.359769909+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq83xlast/model/ with prefix 6aa8eacbc063454a
[INFO 2024-03-28T19:03:42.371659698+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:03:42.371707147+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2024-03-28T19:03:43.544321861+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7jxk6qjh/model/ with prefix 4654e58ed7a64b60
[INFO 2024-03-28T19:03:43.551013252+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2024-03-28T19:03:44.756235664+00:00 kernel.cc:1214] Loading model from path /tmp/tmprvsr8z5p/model/ with prefix c83ecb010074412b
[INFO 2024-03-28T19:03:44.764705897+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2024-03-28T19:03:46.127610072+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk1pvt2q6/model/ with prefix eaddebfbde2344b4
[INFO 2024-03-28T19:03:46.141124995+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2024-03-28T19:03:47.558883539+00:00 kernel.cc:1214] Loading model from path /tmp/tmp95eczq67/model/ with prefix 119bb04d18454f32
[INFO 2024-03-28T19:03:47.573199991+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2024-03-28T19:03:48.981470372+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjyhwwuhi/model/ with prefix 1081fb6a681d4a9c
[INFO 2024-03-28T19:03:48.995506833+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2024-03-28T19:03:50.652972183+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgwls7h1m/model/ with prefix 220e1543b74c4b14
[INFO 2024-03-28T19:03:50.6723198+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2024-03-28T19:03:52.143400266+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcvot8f2f/model/ with prefix 592e9387fc7b443f
[INFO 2024-03-28T19:03:52.160107034+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2024-03-28T19:03:53.614746922+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmwfb2st5/model/ with prefix 0822be212f4a4b05
[INFO 2024-03-28T19:03:53.620470764+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:03:53.62052501+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2024-03-28T19:03:54.861394605+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsbty0987/model/ with prefix b7005830e9b34c86
[INFO 2024-03-28T19:03:54.868156869+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2024-03-28T19:03:56.163356059+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbtv9ir6_/model/ with prefix c9bcd2bbf944488e
[INFO 2024-03-28T19:03:56.172947644+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2024-03-28T19:03:57.955428646+00:00 kernel.cc:1214] Loading model from path /tmp/tmps82gei1g/model/ with prefix 8a9b0735122a45f2
[INFO 2024-03-28T19:03:57.961963781+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2024-03-28T19:03:59.442174007+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfy5ujnyu/model/ with prefix f084ee9b7bb84b82
[INFO 2024-03-28T19:03:59.454630152+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2024-03-28T19:04:01.220475393+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx0ekxlfl/model/ with prefix 3d445704920f4752
[INFO 2024-03-28T19:04:01.231966523+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2024-03-28T19:04:02.565988691+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj20gdjts/model/ with prefix 95b3db24485949e9
[INFO 2024-03-28T19:04:02.573308523+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2024-03-28T19:04:03.816973923+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6_9g9iw7/model/ with prefix 41d55e8450914d9c
[INFO 2024-03-28T19:04:03.82290135+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:04:03.822948473+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2024-03-28T19:04:05.335684684+00:00 kernel.cc:1214] Loading model from path /tmp/tmp617wo8m8/model/ with prefix 6ef935f85f0548de
[INFO 2024-03-28T19:04:05.34981414+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2024-03-28T19:04:07.606739031+00:00 kernel.cc:1214] Loading model from path /tmp/tmpex9gkak_/model/ with prefix b282954f9f88447e
[INFO 2024-03-28T19:04:07.637832664+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2024-03-28T19:04:10.127809113+00:00 kernel.cc:1214] Loading model from path /tmp/tmppa2qny2d/model/ with prefix c3324b569b9d46c6
[INFO 2024-03-28T19:04:10.138388983+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2024-03-28T19:04:12.151329048+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl47ah068/model/ with prefix 0ada537429e64917
[INFO 2024-03-28T19:04:12.158431817+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2024-03-28T19:04:13.69282303+00:00 kernel.cc:1214] Loading model from path /tmp/tmptqxgd2lv/model/ with prefix bc605fbc01214664
[INFO 2024-03-28T19:04:13.70798025+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2024-03-28T19:04:15.058799756+00:00 kernel.cc:1214] Loading model from path /tmp/tmp048w0cfq/model/ with prefix 868354715c1347de
[INFO 2024-03-28T19:04:15.069471422+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:04:15.069541116+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2024-03-28T19:04:16.407974517+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk7ji9_d7/model/ with prefix 54c63d0e49454b8c
[INFO 2024-03-28T19:04:16.417358534+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2024-03-28T19:04:18.028376484+00:00 kernel.cc:1214] Loading model from path /tmp/tmp710dm_nq/model/ with prefix 239994e269354f6d
[INFO 2024-03-28T19:04:18.044665889+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2024-03-28T19:04:19.452692604+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5th9nxd8/model/ with prefix f2fd9997aad14203
[INFO 2024-03-28T19:04:19.462708956+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2024-03-28T19:04:21.037323202+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn3hukwt1/model/ with prefix 1699927915d24416
[INFO 2024-03-28T19:04:21.051943299+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2024-03-28T19:04:22.591667126+00:00 kernel.cc:1214] Loading model from path /tmp/tmputjpk1xb/model/ with prefix 731584bd18b548ef
[INFO 2024-03-28T19:04:22.605319521+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2024-03-28T19:04:23.808003932+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4fql7tp_/model/ with prefix 8e96730f83bf477c
[INFO 2024-03-28T19:04:23.813181139+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2024-03-28T19:04:25.408779069+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4fsibit3/model/ with prefix bd4bf6c80a73434c
[INFO 2024-03-28T19:04:25.425174304+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:04:25.425224042+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2024-03-28T19:04:26.763782696+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy3qi_69v/model/ with prefix c96e0d2d935945e0
[INFO 2024-03-28T19:04:26.773752014+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2024-03-28T19:04:28.895224674+00:00 kernel.cc:1214] Loading model from path /tmp/tmp50ridh0h/model/ with prefix e05b0568207f4da0
[INFO 2024-03-28T19:04:28.914643103+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2024-03-28T19:04:30.334677655+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjnwd9528/model/ with prefix c62bf7c6cb5e4d3b
[INFO 2024-03-28T19:04:30.344845578+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2024-03-28T19:04:31.908937422+00:00 kernel.cc:1214] Loading model from path /tmp/tmphyvjwjv2/model/ with prefix 40612a9cdd724b9e
[INFO 2024-03-28T19:04:31.912642559+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2024-03-28T19:04:33.479144895+00:00 kernel.cc:1214] Loading model from path /tmp/tmpje1ad2nl/model/ with prefix 37c678c9a09f40c6
[INFO 2024-03-28T19:04:33.496505145+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2024-03-28T19:04:35.294998597+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnt4zdtcj/model/ with prefix dde1978cc30e4422
[INFO 2024-03-28T19:04:35.309254012+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2024-03-28T19:04:36.801031739+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2yowl_ya/model/ with prefix 2ef3c46236d74481
[INFO 2024-03-28T19:04:36.806409049+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:04:36.806475392+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2024-03-28T19:04:38.157418178+00:00 kernel.cc:1214] Loading model from path /tmp/tmptnljx0nv/model/ with prefix 253f74088ed7419e
[INFO 2024-03-28T19:04:38.166695006+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2024-03-28T19:04:40.115385272+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnz9n1wwu/model/ with prefix c3fc0d3958cc41e6
[INFO 2024-03-28T19:04:40.128232555+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2024-03-28T19:04:41.823011476+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyy7v2euk/model/ with prefix fd7a8298d3fb4acf
[INFO 2024-03-28T19:04:41.840915019+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2024-03-28T19:04:43.386402651+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo61n_cpn/model/ with prefix 2087a2d2bb8e498f
[INFO 2024-03-28T19:04:43.395743268+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2024-03-28T19:04:45.196649487+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnnfdzdtu/model/ with prefix 3d5881705dfb49f3
[INFO 2024-03-28T19:04:45.204712293+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2024-03-28T19:04:46.410540798+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqcauqvi5/model/ with prefix 9d3db1cde1324fa6
[INFO 2024-03-28T19:04:46.414395984+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2024-03-28T19:04:47.928629131+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8c4lhc6w/model/ with prefix 7de50f5235bc48a4
[INFO 2024-03-28T19:04:47.945380421+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:04:47.945433978+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2024-03-28T19:04:50.34117444+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfbzw1fnr/model/ with prefix df91b1202132494d
[INFO 2024-03-28T19:04:50.356314683+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2024-03-28T19:04:51.640745282+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvx6vrq7y/model/ with prefix 927047fcccbd43ab
[INFO 2024-03-28T19:04:51.645608129+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2024-03-28T19:04:52.914734594+00:00 kernel.cc:1214] Loading model from path /tmp/tmpil3tifkf/model/ with prefix 0c8a41c6551d4340
[INFO 2024-03-28T19:04:52.922715818+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2024-03-28T19:04:54.248987879+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoy8oewaa/model/ with prefix f49db310a0b94238
[INFO 2024-03-28T19:04:54.259767357+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2024-03-28T19:04:56.026043685+00:00 kernel.cc:1214] Loading model from path /tmp/tmpazg2x9vp/model/ with prefix a035f9051ff24935
[INFO 2024-03-28T19:04:56.036590516+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2024-03-28T19:04:57.26281336+00:00 kernel.cc:1214] Loading model from path /tmp/tmp09771so4/model/ with prefix 5ee2f2a6f0b5429a
[INFO 2024-03-28T19:04:57.268359544+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2024-03-28T19:04:59.213150999+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqy5lleus/model/ with prefix fcba21ac34e845ec
[INFO 2024-03-28T19:04:59.242465626+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:04:59.242550024+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2024-03-28T19:05:01.781204414+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmmek57le/model/ with prefix 2ff84bc772304bc6
[INFO 2024-03-28T19:05:01.792094017+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2024-03-28T19:05:03.350360673+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc1tedf6k/model/ with prefix 2b7eefbeb69d4b38
[INFO 2024-03-28T19:05:03.359298722+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2024-03-28T19:05:04.693952022+00:00 kernel.cc:1214] Loading model from path /tmp/tmpeknd33o1/model/ with prefix b194c20b5090416b
[INFO 2024-03-28T19:05:04.69954231+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2024-03-28T19:05:06.001508661+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0q5zh0vt/model/ with prefix c5ded0e298414ccb
[INFO 2024-03-28T19:05:06.008286156+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2024-03-28T19:05:07.655957829+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwk6zrxzl/model/ with prefix 378c159caf9f4f7b
[INFO 2024-03-28T19:05:07.677088117+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2024-03-28T19:05:09.142567421+00:00 kernel.cc:1214] Loading model from path /tmp/tmpep9yuklg/model/ with prefix ffc24d5fe94542f7
[INFO 2024-03-28T19:05:09.156038802+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2024-03-28T19:05:10.42014626+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmu8nii9b/model/ with prefix 5a7c64f33fed4bf6
[INFO 2024-03-28T19:05:10.427288187+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:05:10.427332503+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2024-03-28T19:05:11.710071966+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjpdmna1q/model/ with prefix 2a3def5a92e440b7
[INFO 2024-03-28T19:05:11.719464121+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2024-03-28T19:05:13.011876202+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx0a0fk78/model/ with prefix 48589e770e474fc4
[INFO 2024-03-28T19:05:13.020269536+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2024-03-28T19:05:14.596754939+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7sngleb5/model/ with prefix 1f579b82ad2549f8
[INFO 2024-03-28T19:05:14.614073197+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2024-03-28T19:05:15.917968569+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr9mfjqqz/model/ with prefix 2e86ff056c6a491c
[INFO 2024-03-28T19:05:15.925718102+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2024-03-28T19:05:17.415493929+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc2q53fvm/model/ with prefix 0fa546e3616f482a
[INFO 2024-03-28T19:05:17.427825162+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2024-03-28T19:05:18.780372025+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8ix1pp5t/model/ with prefix 97763c7a568a443a
[INFO 2024-03-28T19:05:18.789115523+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2024-03-28T19:05:20.512733165+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl61varyn/model/ with prefix 11d7aeb6a10c4e83
[INFO 2024-03-28T19:05:20.517681636+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:05:20.517728758+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2024-03-28T19:05:22.342455468+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcil1w4jg/model/ with prefix d16146d2125f4e9f
[INFO 2024-03-28T19:05:22.348362133+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2024-03-28T19:05:24.207058799+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1d3xpq_q/model/ with prefix e462a7c0c2b642c8
[INFO 2024-03-28T19:05:24.214208168+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2024-03-28T19:05:25.557657721+00:00 kernel.cc:1214] Loading model from path /tmp/tmplry6_dyw/model/ with prefix e96c7813d1dc487a
[INFO 2024-03-28T19:05:25.565300674+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2024-03-28T19:05:26.902480307+00:00 kernel.cc:1214] Loading model from path /tmp/tmprdjf5qvs/model/ with prefix 636305118ce14b03
[INFO 2024-03-28T19:05:26.913058204+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2024-03-28T19:05:28.811801982+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2fz241vq/model/ with prefix 637beeb9caad499a
[INFO 2024-03-28T19:05:28.823958069+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2024-03-28T19:05:30.13174352+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr1xb9j58/model/ with prefix d1d5f2c8d5004313
[INFO 2024-03-28T19:05:30.14231788+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2024-03-28T19:05:31.982175932+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkp38rkw9/model/ with prefix d0bbc0f486f444e9
[INFO 2024-03-28T19:05:31.993111919+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:05:31.993177631+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2024-03-28T19:05:35.387495903+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy54nvuro/model/ with prefix 59e16de5e6c247ec
[INFO 2024-03-28T19:05:35.409057447+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2024-03-28T19:05:36.865211473+00:00 kernel.cc:1214] Loading model from path /tmp/tmp80bllzun/model/ with prefix 365bbc5b19444805
[INFO 2024-03-28T19:05:36.874153744+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2024-03-28T19:05:38.998278846+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8e9_iane/model/ with prefix f429ab1e6cf749bb
[INFO 2024-03-28T19:05:39.016519055+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2024-03-28T19:05:40.811286952+00:00 kernel.cc:1214] Loading model from path /tmp/tmpltv6ht1_/model/ with prefix 192047baeaec4e82
[INFO 2024-03-28T19:05:40.832439433+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2024-03-28T19:05:42.342067562+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd4opiza0/model/ with prefix 1e0fc4e8027b438c
[INFO 2024-03-28T19:05:42.355070366+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:05:42.355146886+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2024-03-28T19:05:43.698482202+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_ll2hjtg/model/ with prefix 50a8d9a67c934144
[INFO 2024-03-28T19:05:43.703930305+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2024-03-28T19:05:45.791951869+00:00 kernel.cc:1214] Loading model from path /tmp/tmp33voec5r/model/ with prefix bf0743bbbc65464d
[INFO 2024-03-28T19:05:45.80438546+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2024-03-28T19:05:47.327699056+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkci0t_yi/model/ with prefix 7d1b976ec4cc4d56
[INFO 2024-03-28T19:05:47.335728914+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2024-03-28T19:05:49.562238692+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0y87dva0/model/ with prefix f2583d85dd8d4c7a
[INFO 2024-03-28T19:05:49.582098081+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2024-03-28T19:05:51.101145685+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_zeu88yq/model/ with prefix fe3b9cc64907405a
[INFO 2024-03-28T19:05:51.112451112+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2024-03-28T19:05:52.865114422+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfpxlsx1z/model/ with prefix a5cbf3ada1934fc7
[INFO 2024-03-28T19:05:52.873141848+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-03-28T19:05:52.873209508+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2024-03-28T19:05:54.308298813+00:00 kernel.cc:1214] Loading model from path /tmp/tmpksu_t09y/model/ with prefix 6775d1d2c5d64bba
[INFO 2024-03-28T19:05:54.318480997+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2024-03-28T19:05:55.731804965+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwuh3gibr/model/ with prefix f52990ae26e04dad
[INFO 2024-03-28T19:05:55.7448563+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2024-03-28T19:05:57.419527882+00:00 kernel.cc:1214] Loading model from path /tmp/tmpptv6d35s/model/ with prefix 26f958e0cb46478e
[INFO 2024-03-28T19:05:57.425314672+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2024-03-28T19:05:58.850641717+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4n8hxkig/model/ with prefix f66aca4d56074a2b
[INFO 2024-03-28T19:05:58.859756787+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2024-03-28T19:06:00.583567939+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7f48c43w/model/ with prefix 4d87a3603be94c84
[INFO 2024-03-28T19:06:00.598868088+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 